dali/kernels/shape.h

```diff
@@ -0,0 +1,430 @@
+
+  TensorShape(const std::array<int64_t, ndim> &s) : shape(s) {}
+  // We just want zero-initialization
+  TensorShape() : shape{} {}
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {
+    static_assert(sizeof...(Ts) == ndim - 1, "Number of shapes passed must match ndim");
+  }
+
+  reference operator[](int d) { return shape[d]; }
+  const_reference operator[](int d) const { return shape[d]; }
+
+  iterator begin() noexcept { return shape.begin(); }
+  iterator end() noexcept { return shape.end(); }
+  const_iterator begin() const noexcept { return shape.begin(); }
+  const_iterator end() const noexcept { return shape.end(); }
+  const_iterator cbegin() const noexcept { return shape.cbegin(); }
+  const_iterator cend() const noexcept { return shape.cend(); }
+  reverse_iterator rbegin() noexcept { return shape.rbegin(); }
+  reverse_iterator rend() noexcept { return shape.rend(); }
+  const_reverse_iterator rbegin() const noexcept { return shape.rbegin(); }
+  const_reverse_iterator rend() const noexcept { return shape.rend(); }
+  const_reverse_iterator crbegin() const noexcept { return shape.crbegin(); }
+  const_reverse_iterator crend() const noexcept { return shape.crend(); }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> first() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[i];
+    }
+    return result;
+  }
+
+  template <int other_ndim>
+  TensorShape<other_ndim> last() {
+    static_assert(0 <= other_ndim && other_ndim <= ndim,
+                  "Number of elements in subshape must be between 0 and size()");
+    TensorShape<other_ndim> result;
+    int start_offset = ndim - other_ndim;
+    for (int i = 0; i < other_ndim; i++) {
+      result[i] = (*this)[start_offset + i];
+    }
+    return result;
+  }
+
+  std::array<int64_t, ndim> shape;
+  constexpr int size() const { return ndim; }
+};
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  if (left.size() != right.size()) {
+    return false;
+  }
+  int size = left.size();
+  for (int i = 0; i < size; i++) {
+    if (left[i] != right[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator!=(const TensorShape<left_ndim> &left, const TensorShape<right_ndim> &right) {
+  return !(left == right);
+}
+
+template <int left_ndim, int right_ndim>
+typename std::enable_if<left_ndim != DynamicDimensions && right_ndim != DynamicDimensions, 
```

Comment:
Again, I'd go for a `cat`.

Reply:
```
constexpr int AddShapeDim(int d1, int d2) { return d1 == DynamicDimensionss || d2 == DynamicDimensions ? DynamicDimensions : d1+d2; }

template <int d1, int d2>
TensorShape<AddShapeDim(d1, d2)> cat(const TensorShape<d1> &s1, const TensorShape<d2> &s2) {
  auto out_dim = AddShapeDim(d1, d2);
  TensorShape<out_dim> out;
  if (out_dim != DynamicDimensions) // avoid unnecessary check in resize
    out.resize(d1+d2);
  // run the loops here...
}
```

Reply:
Done, although I don't like the idea of mixing the static shapes with dynamic shapes.